fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -484,7 +508,7 @@ private static void validateRecords(@Nonnull Descriptors.FileDescriptor fileDesc
         }
     }
 
-    private Descriptors.Descriptor initRecordTypes(boolean processExtensionOptions) {
+    private Descriptors.Descriptor initRecordTypes(@Nullable RecordMetaData evolvedRecordMetaData, boolean processExtensionOptions) {
```

Comment:
Discussion copied from #276:

@alecgrieser:

> I think generated the record meta-data like this is a little too heavy weight. If I'm reading the code correctly, the reason for passing it here (rather than just passing the evolved descriptor) is to get the union descriptor. Would it be cleaner to refactor that code out of this method, and then this method only needs the evolved descriptor (and not the full meta-data)?

@amouehsan:

> It's also used for record types. The filedescriptor only gives a list of descriptors, not a map of them. I was initially using a RecordMetaDataBuilder, but while they are not very different, this felt more natural. I can create the builder / metadata inside this method if you prefer so.

Reply:
I still feel like creating the record meta-data from the evolved descriptor is too heavy weight just to get (1) the union descriptor message and (2) the map from record type names to record types (especially as all that is needed from the record type is the descriptor.

I think once you've extracted the union descriptor, you should be able to extract a mapping from type names to descriptors from that. (And if the type is missing in the union descriptor, that it's own error.)

Reply:
I may be wrong, but I think the union descriptor only has a subset of record types; I.e., not all record types are necessarily present in the union. Assuming this is true, if we do not want to create a builder, we probably have two options: 
(a) Get the updated descriptors only if they are present in the union. Would this be correct? it would solve the problem in issue #3, but not sure if it's correct from the schema evolution point of view; some descriptors will come from the local descriptor and some will come from the proto.
(b) We could scan the list of message types in the evolved file descriptor once, and create a map. Let me give the second a shot.

Reply:
OK, (a) doesn't seem correct to me. Also,  I gave (b) a shot but it is not going to work if we want to support this:

> A record type is moved from the descriptor to being in an imported file in the evolved descriptor. Then the record should be present (using the imported file) in the new meta-data.

The record types are not in the messageTypes list, and we need to fish it from the imported files.

Extracting the descriptors from the union descriptor without verifying that the record type is already known, does not look right to me either. 

To fetch all of the record types defined in the local file descriptor, we may need to do most of the heavy work that the builder does. Hence, IMO, the cleanest way of doing this is to build the builder.

Reply:
Maybe I'm missing something.

By contract, all record types *must* be defined in the union descriptor (until such time as we decide we generate the union descriptor from the records, but we're not there yet). So if you look at all record types defined the records descriptor's union descriptor, they should each exist in the *local* file descriptor's union descriptor (though they may be imported in one descriptor or the other).

> Extracting the descriptors from the union descriptor without verifying that the record type is already known, does not look right to me either.

This is essentially what the logic to handle importing records from other files is already doing, right?

Reply:
Iterated thru the list of message types and union fields, as a superset of all record types. done.